### ISSUE TITLE
[Question] Trying to understand and implement redux-orm

### ISSUE THREAD

**Issue by nealoke** [tommikaikkonen/redux-orm#106]
I'm trying to understand how this works with [the example provided ](https://github.com/tommikaikkonen/redux-orm-primer/tree/migrate_to_0_9) but I find it quite difficult to understand this concept and library. I'm getting really confused...

**Question 1:**
The `Todo` model field declarations states `tags: many('Tag', 'todos')`. What I can't understand is where the `'todos'` comes from and where it ends up coming from. Please explain :( 

**Question 2:**
How would you use orm together with normalizr?

**Question 3:**
What are sessions, and why are they so important to orm? Why would you want more then 1 session used in your application?

**Comment by markerikson** [comment]
1) The `"todos"` argument is used to define the generated field name on the Tag side of the relationship.  So, each Tag model instance will have a field `todos` to represent all the Todos that Tag is associated with.

2) If you're using Redux-ORM, you generally won't need Normalizr.

3) A "session" wraps up operations on a specific JS object with the "tables" full of data.  You would create a session each time you need to do work based on a specific version of your app state.

I wrote a pair of blog posts that describe basic and advanced use of Redux-ORM: [Practical Redux, Part 1: Redux-ORM Basics](http://blog.isquaredsoftware.com/2016/10/practical-redux-part-1-redux-orm-basics/) and [Practical Redux, Part 2: Redux-ORM Concepts and Techniques](http://blog.isquaredsoftware.com/2016/10/practical-redux-part-2-redux-orm-concepts-and-techniques/).  They cover Redux-ORM 0.8, not the almost-finished 0.9, but most of the ideas are still the same.

**Comment by nealoke** [comment]
@markerikson thanks for helping so fast! And yes I've been looking at these posts as well. The only thing that is not so clear to me (because I don't fully understand it) is how you would switch from a normalized state like below to a omr model. Would love some help :)

```
users: {
	results: ['user hash 1'],
	entities: {
		'user hash 1': {
			firstName: 'Neal',
			lastName: 'Van der Valk',
			email: 'email@outlook.com', 

			notifications: {
				'daily': true,
				'weekly': false
			},

			permissions: {
				'admin': true,
				'results': true,
				'database': true,
				'download': true,
				'account': true,
				'departments': true,
				'users': true,
				'devices': true,	
				'integrations': true,
			},

			departments: ['dept hash 1', 'dept hash 2']
		} 
	}
}
```

**Comment by markerikson** [comment]
If you read the rest of my "Practical Redux" series, you can see how I load data from a request into the store via Redux-ORM, at which point Redux-ORM handles the normalization part.  (I _think_ that's what you're asking, but maybe I'm wrong.)

**Comment by nealoke** [comment]
@markerikson aah didn't see that it went on about orm after part 2, great! Another question, this time about the `querySet`. When would you use `toRefArray()` and when `toModelArray()`?

**Comment by chanan** [comment]
`toRefArray()` returns the plain javascript object, so if you have fk or many fields they will be ids. so for example `user.deparment === 1`

`toModelArray()` returns the "enhanced" javascript object with redux-orm methods. So if you have a fk or many it would actually be the "real thing" for example `user.department.name` will work

**Comment by nealoke** [comment]
@chanan thank you, I've come a long way since yesterday!

**Comment by nealoke** [comment]
Thank you guys for all the help, I've come to implement it fully now! 👍 

Another question that came up was if there is a way to reduce the following boilerplate, as you would write this in all your selectors. Maybe with wrapping function or so?

```
import orm from '../orm';
import { createSelector } from 'reselect';
import { createSelector as ormCreateSelector } from 'redux-orm';

// Selects the state managed by Redux-ORM.
export const ormSelector = state => state.orm;

export const users = createSelector(
	ormSelector,
	ormCreateSelector(orm, (session) => {
		return session.user.all().toModelArray().map(user => {
			const obj = Object.assign({}, user.ref);
			return obj;
		});
	})
);
```

**Comment by chanan** [comment]
You could just write:

`return session.User.all().toRefArray()`

**Comment by nealoke** [comment]
@chanan cool that cut's it down a bit but I was thinking about the following part, but maybe i'm asking to much here :)

```
export const users = createSelector(
	ormSelector,
	ormCreateSelector(orm, (session) => { /* ... */});
```

**Comment by nealoke** [comment]
@chanan I'm not sure why but it looks like i'm the only one that finds the docs super confusing as parts are up to date and parts aren't ([even in the example which should be up to date](https://github.com/tommikaikkonen/redux-orm-primer/tree/migrate_to_0_9)). I'm currently trying to intialise the orm and put an initial record in there. currently with the following code, but it isn't working. 

The `dbSession.session().create(/* ... */)` is working fine as I can see in `dbSession.state` afterwards, but it is not updating the actual orm state in redux.

```
import { ORM } from 'redux-orm';
import { session, user, department, account, question, questionAnswer, device } from './models';

const orm = new ORM();
orm.register(session, user, department, account, question, questionAnswer, device);

// Creating initial session for login
const dbSession = orm.session(orm.getEmptyState());
dbSession.session.create({
	needsLogin: true,
	isFetching: false,
	loginError: false,
	user: '',
	account: ''
});

export default orm;
```

**Comment by markerikson** [comment]
@nealoke : that's correct.  In 0.8, calling `session.reduce()` produces a new plain JS data object, which then needs to be returned from a Redux reducer to actually update the store.  In 0.9, an `orm` instance automatically updates its internal `this.state` object with each ORM update step, but you still need to actually tell Redux to use that new data object, again by usually returning it from a reducer.

I'd suggest you re-read "Practical Redux Part 1" to to see the two basic methods of using Redux-ORM as part of reducer logic.

**Comment by nealoke** [comment]
@markerikson so there is not a direct way via orm of telling redux that the new state is present? Also how would implementing this into a reducer solve my issue with providing an initial state? As far as I can see in your blogpost you still need to call  `schema.getDefaultState()`? Or am I missing something?

```
const initialState = schema.getDefaultState();
export default function entitiesReducer(state = initialState, action) {/* ... */}
```

I really thank you for your time and effort!

**Comment by chanan** [comment]
I think you might be confusing Mark's blog post which are for version 0.8 and the current version which 0.9. You do not need to notify the store of new state in 0.9, it is done automatically by the Orm. You might want to peek at: https://github.com/carlesnunez/react-native-weather-app which is an example app using 0.9

**Comment by nealoke** [comment]
@chanan I've seen that it is version 0.8, but still helped me a bit. Thanks for providing the example using 0.9, although I don't see any initial state inside the orm there :(

**Comment by chanan** [comment]
What do you mean? Are you asking how to bootstrap using version 0.9?

**Comment by markerikson** [comment]
With the caveat that I haven't used 0.9 yet (or actually looked fully into how the API changes work), I _think_ these two functions are roughly equivalent:

```js
// 0.8
function someEntitiesReducer(entitiesState, action) {
    const session = schema.from(entitiesState);
    const {SomeModel} = session.
    
    const modelInstance = SomeModel.withId(action.id);
    modelInstance.someField = action.newValue;
    
    const newState = session.reduce();
    return newState;
}


// 0.9
function someEntitiesReducer(entitiesState, action) {
    const session = orm.session(entitiesState);
    const {SomeModel} = session;
    
    const modelInstance = SomeModel.withId(action.id);
    modelInstance.someField = action.newValue;
    
    // Redux-ORM automatically created a new version of session.state, so we just grab it
    const newState = session.state;
    return newState
}
```

**Comment by nealoke** [comment]
@chanan exactly

**Comment by chanan** [comment]
Here is some of my bootstrap code for 0.9:

```
export default function bootstrap(orm) {
    const emptyDBState = orm.getEmptyState()
    const session = orm.session(emptyDBState)
    const { Item, Section, Form, Test } = session

    //*******************************
    // Tests
    //*******************************
    const tests = { }
    tests[5] = Test.create({
        id: 5,
        name: 'SEU Placement'
    })

//CUT

//*******************************
    // Forms
    //*******************************
    const forms = { }
    forms[0] = Form.create({
        name: 'SEU Official',
        numberOfItems: 0,
        creationDate: '03/31/16',
        modificationDate: '01/21/17',
        formType: 'Linear',
        language: 'ARABIC',
        duration: 2,
        time: {
            minutes: 15,
            seconds: 30
        },
        test: tests[5]
    })

//CUT

```

You call it like so:

```
const store = createStoreWithMiddleware(reducer, bootstrap(orm), window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__())
```

**Comment by markerikson** [comment]
Oh, bootstrapping, not _just_ updates.

If you only need the empty "tables", you can use `schema.getDefaultState()` in 0.8, or `orm.getEmptyState()` in 0.9.

If you want that data _populated_, you can write an initialization function like this:

```js
// 0.8
function createInitialState() {
    const session = schema.from(schema.getDefaultState());
    const {SomeModel} = session.
    
    const modelInstance = SomeModel.withId(action.id);
    modelInstance.create({id : 1, name "Blah"});
    
    const newState = session.reduce();
    return newState;
}



// 0.9
function createInitialState() {
    const session = orm.session(orm.getEmptyState());
    const {SomeModel} = session;
    
    const modelInstance = SomeModel.withId(action.id);
    modelInstance.create({id : 1, name "Blah"});
    
    // Redux-ORM automatically created a new version of session.state, so we just grab it
    const newState = session.state;
    return newState
}


// Either way, you can then do:

const state = createInitialState();

function entitiesReducer(state = initialState, action) {
    // usual reducer logic here
}
```

**Comment by nealoke** [comment]
@chanan do I also return something at the end?

**Comment by chanan** [comment]
Sorry, edited the comment to add the return.

**Comment by nealoke** [comment]
@chanan I'm not sure but I don't see it :), for clarification I mean do I return an object or so in the `bootstrap` function.

**Comment by chanan** [comment]
Yes I updated the example above. You might need to refresh this page. But just in case I am pasting it here as well:

```
return {
        orm: session.state
    }
```

I also added a root reducers snippet to the example above. Again pasting as well:

```
import { combineReducers } from 'redux'
import { createReducer } from 'redux-orm'

import { orm } from '../models'


const app= combineReducers({
    orm: createReducer(orm)
})

export default app
```

**Comment by nealoke** [comment]
@chanan yes just saw the update <3, what would I do without you! Also @markerikson thank you!

**Comment by plandem** [comment]
it's better to use this guide as reference:
https://github.com/tommikaikkonen/redux-orm/wiki/0.9-Migration-Guide

**Comment by nealoke** [comment]
@chanan @plandem  @markerikson  I'm back with another question. Up untill now i've always followed the example etc to implement the `static reducer()` in the orm model. How would I go about to break it away to it's own file and "real" reducer?

I understand the part on how to make the reducer (as shown below), but I can't seem to find how to fit the created reducer back into the orm model / reducer. (would love to keep everything from the orm under the orm reducer).

**Example reducer file**
```
import { orm } from './models';

// Let's say in this example that this would be the reducer for
// the "book" model provided in the docs.

function bookReducer(dbState, action) {
    const sess = orm.session(dbState);
    const { Book } = sess;

    switch (action.type) {
        // Cases go here
    }

    return sess.state;
}

// Where will this function be implemented ?
```


**Rootreducer file**
```
import { combineReducers } from 'redux';

// ORM
import { createReducer } from 'redux-orm';
import orm from '../initializers/orm';

// Reducers
import { reducer as formReducer } from 'redux-form';
import { reducer as notificationsReducer } from 'react-notification-system-redux';
import popupReducer from './popup';
import sessionReducer from './session';

const rootReducer = combineReducers({
	session: sessionReducer,
	popup: popupReducer,
	orm: createReducer(orm), // Would like to keep it like it is here!
	form: formReducer,
	notifications: notificationsReducer
});

export default rootReducer;
```

**Comment by plandem** [comment]
almost as before:
```
class Book extends Model {
    static reducer(action, Book, session) {
....
    }
}
```

But:
1. you don't need to return anything now
2. if you don't use built-in reducer (created with **createReducer**), then you have to re-implement logic to call model's reducer. Here is built-in implementation:
```
function createReducer(orm, updater = defaultUpdater) {
    return (state, action) => {
        const session = orm.session(state || orm.getEmptyState());
        updater(session, action);
        return session.state;
    };
}

function defaultUpdater(session, action) {
    session.sessionBoundModels.forEach(modelClass => {
        if (typeof modelClass.reducer === 'function') {
            modelClass.reducer(action, modelClass, session);
        }
    });
}
```

As you can see, **defaultUpdater** calls reducer of each model

**Comment by nealoke** [comment]
@plandem thank you so much! I almost had it figured out but you helped the last bit! 👍. By now we could almost  have the entire docs rewritten 🤣

**Comment by nealoke** [comment]
@plandem sorry to keep bothering you :D 

I'm looking to create a selector as done in the docs like so:
```
// (re)selector
import orm from '../initializers/orm';
import { createSelector } from 'reselect';
import { createSelector as ormCreateSelector } from 'redux-orm';

/*----------  Select all patients  ----------*/
export const patientsSelector = createSelector(state => state.orm, ormCreateSelector(orm, (session) => {
	return session.Patient.all().toRefArray();
}));
```

However when implementing it in the `mapStateToProps` it doesn't see the changed `state.orm` when a value gets added after the initial render of the component.

**In simple words: this doesn't work**
```
// In component
function mapStateToProps(state) {
    return {
        authors: authorSelector(state.orm),
    };
}
```

**This does work**
```
// In component
function mapStateToProps(state) {
    return {
        authors: authorSelector(state), // so not state.orm
    };
}
```

Am I doing something wrong here?

**Comment by chanan** [comment]
Nope, you are correct. The reason is, look at the first param of your selector. it takes state and returns state.orm
```
 state => state.orm
```

**Comment by nealoke** [comment]
@chanan Aah I see! Thanks!

**Comment by nealoke** [comment]
@chanan is it also possible to pass arguments to a selector like so? If so, how do I structure the selector, or do I just simply 

```
const mapStateToProps = (state) => ({
	patient: patientSelector(state, 'another parameter'),
	consultations: '',
	therapies: ''
})
```

**This isn't working, i'm not sure where the parameter would end up**
```
export const patientSelector = createSelector(state => state.orm, ormCreateSelector(orm, something, (session) => {
	console.log(something);
	return session.Patient.all().toRefArray();
}));
```

**Comment by chanan** [comment]
Here is a working example from my code that might help you. I should warn you, I am not an expert in JS and I don't really use reselect anywhere else, so I might be doing this "wrong"

```
const ormSelector = state => state.orm
const propsSelector = (state, props) => props
const idSelector = (state, props) => props.params !== undefined ? props.params.id : props.id

export const getTest = createSelector(orm, ormSelector, propsSelector, idSelector, (session, props, id) =>
    session.Test.withId(id).ref
)
```

**Comment by nealoke** [comment]
@chanan you are an expert to me! :D Could you maybe also show how you call it in the component etc?

**Comment by chanan** [comment]
No problem:

```
const mapStateToProps = (state, props) => ({
    test: getTest(state, props)
})

```

**Comment by nealoke** [comment]
@chanan I see that you don't use the `ormCreateSelector` from orm, is there a reason, or is it just not necessary to also use this with reselect?

**Comment by chanan** [comment]
Good question and to be honest I don't recall which is 0.9 way of doing things. I can tell you that I think its because I copied the pattern from here:

https://github.com/carlesnunez/react-native-weather-app/blob/master/src/selectors/citySelector.js

But I really don't remember.

**Comment by nealoke** [comment]
@chanan I guess so then, will remove it for now and stick with your example. I'm not sure how I can thank you, but damn thank you so much :)

**Comment by chanan** [comment]
no worries!

**Comment by nealoke** [comment]
@chanan @markerikson 

can I ask you, how would you manage data that comes in dynamically? So bascially if I'm on a page I then only get the data related data.

My example is this, when on a patient's profile, I want to fetch all their previous consulations. I currently have an action creator `fetchPatientConsultations` which get's all consultations, but ideally I would have an action creator `fetchConsulation` which only get's the consultations not yet in the orm db.

How would I access the patient's consultations if `session.Patient.withId(hash).consultations.toRefArray()` in my selector always returns `[]` (because it is not in the database yet). It should be accessible somewhere because on my `Patient` model it is supplied by the initial fetch.

**I want to access this!** 
![image](https://cloud.githubusercontent.com/assets/17417414/24579699/7c8a8416-16fa-11e7-9754-4c18d457369f.png)

**Comment by chanan** [comment]
Sorry, I am not %100 following the problem. I will start by saying that I am not fetching anything yet myself, I am working on the app fully offline at this time. However, I know I will be using redux-saga to fetch data. Basically, in the Saga get your orm and find out what is in it that is missing and do a call to the server to get those items.

**Comment by nealoke** [comment]
@chanan and how would you get your orm and query it in a saga then? :)

**Comment by chanan** [comment]
I don't know but it seems to work :)

For example in one of my sagas I need to get an item for some reason or other and this works:

```
const item = yield select(getItemWithSections, { params: { id }})
```

That params you see there is just because of my idSelector I should you earlier.

**Comment by nealoke** [comment]
@chanan I'm not really into sagas but more thunks, but i'm sure it is not that different. Do you mind explaining what the `select()` function and `getItemWithSections` are?

**Comment by chanan** [comment]
getItemWithSections is my orm selector, so that is easy to understand just for completeness I am including it below, replace that call with your selector.

```
export const getItemWithSections = createSelector(orm, ormSelector, propsSelector, idSelector, (session, props, id) => {
    const item = session.Item.withId(id)
    return Object.assign({}, item.ref, {
        sections: item.sections.toRefArray().map(section => section.id)
    })
})
```

`select()` is a saga function, it calls my selector and I am guessing it passes state to it. Here is the doc for it: https://redux-saga.github.io/redux-saga/docs/api/index.html#selectselector-args

Sorry, never used thunk so I cant tell you the equivalent way of doing it using thunk.

**Comment by nealoke** [comment]
@chanan Just to recap, you always use selectors when going to the orm db then right?

Also when you do ` sections: item.sections.toRefArray().map(section => section.id)`, does this also work when there is not data in sections yet?

AAH i'm pulling my hair out on how to explain the issue xD, let me walk you through the process in my case.

**Step 1: Receiving the initial data**
In this step I receive the individual patients. The patient linked fields are as shown below. Please note that while I receive a patient, I don't receive the data of `consultations` or `therapies`. This means that in my orm db I receive `patients` but it strips away the linked fields to some place else.
```
// Model fields
Patient.fields = {
	consultations: many({
		to: 'Consultation',
		relatedName: 'consultations'
	}),

	therapies: many({
		to: 'Therapy',
		relatedName: 'therapies'
	})
};

// Example of received patient
{
	// Other stuff
	consultations: ['consultationHash'],
	therapies: ['therapyHash']
}
```

Where it stores the linked fields (same goes for therapies).

![image](https://cloud.githubusercontent.com/assets/17417414/24617669/e3a68276-1894-11e7-827d-ada20be93eb2.png)



**Step 2: query db to figure out which consultations / therapies to fetch**
By now we have patients in the db. When viewing a patient's profile, I want to fetch their consultations / therapies. To do this I tried it with the following selector, however it always returns an empty array. Even though that on step 1, the linked fields were processed by the orm db (see picture above). I can't find a way to query these linked fields.

```
/*----------  Patient's consultations  ----------*/
export const patientConsultationsSelector = createSelector(orm, ormSelector, hashSelector, (session, hash) => {
	return session.Patient.hasId(hash) ? session.Patient.withId(hash).consultations.toRefArray() : null;
});

// Returns [] even though I can view the linked field via redux inspection tools (picture above)

```

It is my intention to be able to call the above selector, get all the id's of the linked fields and then call an action creator which fetches the data for that id for example `fetchConsultation(id)`.

Is this clearer? If not please ask, you are helping me a lot!

**Comment by chanan** [comment]
I am still not 100% sure I understand but I am going to take a stab at it. Are you getting the Ids but not the object, then saving them in your many to many field but the "real objects" aren't there? If so, I am *guessing* that maybe the orm is getting confused because it cant find the objects and is returning null. If you think that is the case try doing the following:

1. Add a field to your object that is not a relationship, but stores you ids
2. In your selector, read that field instead
3. Load your consultations & therapies based on that new field(s) you just added.

Was I on the mark on that one?

**Comment by nealoke** [comment]
@chanan Yep that is exactly what I'm trying to say :D. The orm sees the connection fields, but the objects are not there yet.

Although I could totally do what you are recommending, I still feel that this functionality should be inside the library though. Especially when focusing on react/redux where data comes and goes async. 

The "annoying" thing is that you can see in the orm db that it stores the connection as shown on the picture above even though the object is not present, but there is simply no way of retrieving it out of the db i guess :(.

**Comment by chanan** [comment]
Well, I didn't write the library :)

But I actually see the reasoning there. It's an orm, so it deals with objects and in a sense you are trying to "trick it", so that it sometimes deals with ids and sometimes with objects. So, I think you should try what I suggested and make sure it works. And after that, if you feel strongly about it, submit a PR that makes it work both ways (should be possible), and see if the creator of the library accepts it.

**Comment by nealoke** [comment]
@chanan I know I know ;)

I'll try the implementation you suggested! To bad the author of this library is not really active anymore :(

thanks again for the great help!

**Comment by carlesnunez** [comment]
@nealoke Hi, I am the owner of the https://github.com/carlesnunez/react-native-weather-app example

Do u still having the issue?

Explain me your problem in a few words and i'll try to help u :)

**Comment by nealoke** [comment]
@carlesnunez I found a work around for this issue but I think it's not optimal though. 

Currently I insert empty objects with default fields like `didFetch: false` in the relational model which data is not yet fetched. When I need the data I check if `didFetch` is false, if so get the data.

**Comment by fabdarice** [comment]
@carlesnunez Hi, are you available to chat? I have a few questions regarding ORM that I'm having trouble understanding.

**Comment by carlesnunez** [comment]
Almost on bed, here in spain is about 01 AM but text me here your question and i'll help you for sure tomorrow!

**Comment by carlesnunez** [comment]
@fabdarice

**Comment by fabdarice** [comment]
@carlesnunez, thanks I appreciate your help !
Here's a ticket describing the concept I'm really struggling with : 
https://github.com/tommikaikkonen/redux-orm/issues/150

Feel free to reach out to me if you don't understand what are my questions.

**Comment by carlesnunez** [comment]
You have a response on your issue @fabdarice
